File: app/src/main/java/ru/forblitz/statistics/widget/data/PlayerFastStat.java
package ru.forblitz.statistics.widget.data;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;

import ru.forblitz.statistics.R;
import ru.forblitz.statistics.data.Constants;
import ru.forblitz.statistics.data.StatisticsData;
import ru.forblitz.statistics.utils.InterfaceUtils;

public class PlayerFastStat extends LinearLayout {

    public PlayerFastStat(Context context) {
        super(context);
    }

    public PlayerFastStat(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public PlayerFastStat(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public void setData(StatisticsData statisticsData) {

        TextView battles = this.findViewWithTag("battles");
        TextView winRate = this.findViewWithTag("winRate");
        TextView averageDamage = this.findViewWithTag("averageDamage");
        TextView efficiency = this.findViewWithTag("efficiency");
        TextView survived = this.findViewWithTag("survived");
        TextView hitsFromShots = this.findViewWithTag("hitsFromShots");

        String battlesValue = statisticsData.getBattles();
        String winRateValue = statisticsData.getWinRate();
        String averageDamageValue = statisticsData.getAverageDamage();
        String efficiencyValue = statisticsData.getEfficiency();
        String survivedValue = statisticsData.getSurvived();
        String hitsFromShotsValue = statisticsData.getHitsFromShots();

        int winRateColor = InterfaceUtils.getValueColor(this.getContext(), statisticsData.getWinRate(), Constants.Steps.winRate);
        int averageDamageColor = InterfaceUtils.getValueColor(this.getContext(), statisticsData.getAverageDamage(), Constants.Steps.averageDamage);
        int efficiencyColor = InterfaceUtils.getValueColor(this.getContext(), statisticsData.getEfficiency(), Constants.Steps.efficiency);

        if (winRateValue.equals("NaN")) { winRateValue = getContext().getString(R.string.empty); winRateColor = this.getContext().getColor(R.color.white); }
        if (averageDamageValue.equals("NaN")) { averageDamageValue = getContext().getString(R.string.empty); averageDamageColor = this.getContext().getColor(R.color.white); }
        if (efficiencyValue.equals("NaN")) { efficiencyValue = getContext().getString(R.string.empty); efficiencyColor = this.getContext().getColor(R.color.white); }
        if (survivedValue.equals("NaN")) { survivedValue = getContext().getString(R.string.empty); }
        if (hitsFromShotsValue.equals("NaN")) { hitsFromShotsValue = getContext().getString(R.string.empty); }

        battles.setText(battlesValue);
        winRate.setText(winRateValue);
        averageDamage.setText(averageDamageValue);
        efficiency.setText(efficiencyValue);
        survived.setText(survivedValue);
        hitsFromShots.setText(hitsFromShotsValue);

        battles.setTextColor(InterfaceUtils.getValueColor(this.getContext(), statisticsData.getBattles(), Constants.Steps.battles));
        winRate.setTextColor(winRateColor);
        averageDamage.setTextColor(averageDamageColor);
        efficiency.setTextColor(efficiencyColor);

    }

    public void setSessionData(StatisticsData statisticsData) {

        String battlesDiff = statisticsData.getBattles();
        String winRateDiff = statisticsData.getWinRate();
        String averageDamageDiff = statisticsData.getAverageDamage();
        String efficiencyDiff = statisticsData.getEfficiency();
        String survivedDiff = statisticsData.getSurvived();
        String hitsFromShotsDiff = statisticsData.getHitsFromShots();

        battlesDiff = "+" + battlesDiff;
        if (Double.parseDouble(winRateDiff) > 0) { winRateDiff = "+" + winRateDiff; }
        winRateDiff = winRateDiff + "%";
        if (Double.parseDouble(averageDamageDiff) > 0) { averageDamageDiff = "+" + averageDamageDiff; }
        if (Double.parseDouble(efficiencyDiff) > 0) { efficiencyDiff = "+" + efficiencyDiff; }
        if (Double.parseDouble(survivedDiff) > 0) { survivedDiff = "+" + survivedDiff; }
        survivedDiff = survivedDiff + "%";
        if (Double.parseDouble(hitsFromShotsDiff) > 0) { hitsFromShotsDiff = "+" + hitsFromShotsDiff; }
        hitsFromShotsDiff = hitsFromShotsDiff + "%";

        ((DifferenceIndicatorView) this.findViewWithTag("battlesDiff")).setValue(battlesDiff, false);
        ((DifferenceIndicatorView) this.findViewWithTag("winRateDiff")).setValue(winRateDiff, true);
        ((DifferenceIndicatorView) this.findViewWithTag("averageDamageDiff")).setValue(averageDamageDiff, false);
        ((DifferenceIndicatorView) this.findViewWithTag("efficiencyDiff")).setValue(efficiencyDiff, false);
        ((DifferenceIndicatorView) this.findViewWithTag("survivedDiff")).setValue(survivedDiff, true);
        ((DifferenceIndicatorView) this.findViewWithTag("hitsFromShotsDiff")).setValue(hitsFromShotsDiff, true);

    }

}

File: app/src/main/java/ru/forblitz/statistics/widget/data/PlayerFastStat.kt
package ru.forblitz.statistics.widget.data

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import ru.forblitz.statistics.ForBlitzStatisticsApplication
import ru.forblitz.statistics.R
import ru.forblitz.statistics.data.Constants
import ru.forblitz.statistics.data.Constants.PreferencesSwitchesTags
import ru.forblitz.statistics.dto.StatisticsData
import ru.forblitz.statistics.utils.InterfaceUtils
import ru.forblitz.statistics.utils.ParseUtils
import ru.forblitz.statistics.utils.RoundingUtils.Companion.round

class PlayerFastStat : LinearLayout {
    constructor(context: Context?) : super(context)
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )

    private val averageDamageRounding: Boolean
        get() = (context.applicationContext as ForBlitzStatisticsApplication).preferencesService.get(PreferencesSwitchesTags.averageDamageRounding)

    @SuppressLint("SetTextI18n")
    fun setData(nickname: String, statisticsData: StatisticsData) {
        with(findViewWithTag<TextView>("battles")) {
            text = ParseUtils.splitByThousands(statisticsData.battles.toString())
            setTextColor(
                InterfaceUtils.getValueColor(
                    this.context,
                    statisticsData.battles.toDouble(),
                    Constants.Steps.battles
                )
            )
        }
        with(findViewWithTag<TextView>("winRate")) {
            if (statisticsData.winningPercentage!!.isNaN()) {
                text = context.getString(R.string.empty)
                setTextColor(context.getColor(R.color.white))
            } else {
                text = "${statisticsData.winningPercentage!!.round()}%"
                setTextColor(InterfaceUtils.getValueColor(
                    context,
                    statisticsData.winningPercentage!!.toDouble(),
                    Constants.Steps.winRate
                ))
            }
        }
        with(findViewWithTag<TextView>("averageDamage")) {
            if (statisticsData.averageDamage!!.isNaN()) {
                text = context.getString(R.string.empty)
                setTextColor(context.getColor(R.color.white))
            } else {
                text = ParseUtils.splitByThousands(
                    if (!averageDamageRounding) {
                        statisticsData.averageDamage!!.toInt().toString()
                    } else {
                        statisticsData.averageDamage!!.round().toString()
                    }
                )
                setTextColor(InterfaceUtils.getValueColor(
                    this.context,
                    statisticsData.averageDamage!!.toDouble(),
                    Constants.Steps.averageDamage
                ))
            }
        }
        with(findViewWithTag<TextView>("efficiency")) {
            if (statisticsData.efficiency!!.isNaN()) {
                text = context.getString(R.string.empty)
                setTextColor(context.getColor(R.color.white))
            } else {
                text = statisticsData.efficiency!!.round().toString()
                setTextColor(InterfaceUtils.getValueColor(
                    this.context,
                    statisticsData.efficiency!!.toDouble(),
                    Constants.Steps.efficiency
                ))
            }
        }
        with(findViewWithTag<TextView>("survived")) {
            text = if (statisticsData.survivedPercentage!!.isNaN()) {
                context.getString(R.string.empty)
            } else {
                statisticsData.survivedPercentage!!.round().toString()
            }
        }
        with(findViewWithTag<TextView>("hitsFromShots")) {
            text = if (statisticsData.hitsOutOfShots!!.isNaN()) {
                context.getString(R.string.empty)
            } else {
                statisticsData.hitsOutOfShots!!.round().toString()
            }
        }
        findViewWithTag<TextView>("nickname").text = nickname
    }

    fun setSessionData(statisticsData: StatisticsData) {
        (findViewWithTag<View>("battlesDiff") as DifferenceIndicatorView).setValue(
            statisticsData.battles,
            false
        )
        (findViewWithTag<View>("winRateDiff") as DifferenceIndicatorView).setValue(
            statisticsData.winningPercentage!!,
            true
        )
        (findViewWithTag<View>("averageDamageDiff") as DifferenceIndicatorView).setValue(
            if (!averageDamageRounding) {
                statisticsData.averageDamage!!.toInt().toFloat()
            } else {
                statisticsData.averageDamage!!
            },
            false
        )
        (findViewWithTag<View>("efficiencyDiff") as DifferenceIndicatorView).setValue(
            statisticsData.efficiency!!,
            false
        )
        (findViewWithTag<View>("survivedDiff") as DifferenceIndicatorView).setValue(
            statisticsData.survivedPercentage!!,
            true
        )
        (findViewWithTag<View>("hitsFromShotsDiff") as DifferenceIndicatorView).setValue(
            statisticsData.hitsOutOfShots!!,
            true
        )
    }
}